File: src/main/java/com/squarepolka/digiscope/plot/plotpoint/PlotPointPulse.java
package com.squarepolka.digiscope.plot.plotpoint;

import java.math.BigDecimal;

public class PlotPointPulse extends PlotPointRaw {

    private PlotPointRaw startPulse;
    private PlotPointRaw endPulse;
    private BigDecimal durationSeconds;

    public PlotPointPulse(PlotPointRaw startPulse, PlotPointRaw endPulse) {
        super(startPulse.getTimestampSeconds(), startPulse.getVoltValue());
        this.startPulse = startPulse;
        this.endPulse = endPulse;
        this.durationSeconds = endPulse.getTimestampSeconds().subtract(startPulse.getTimestampSeconds());
    }


    public BigDecimal getDurationSeconds() {
        return durationSeconds;
    }

    public BigDecimal getDurationMilliseconds() {
        return durationSeconds.multiply(new BigDecimal(1000));
    }

    public BigDecimal getDurationMicroseconds() {
        return durationSeconds.multiply(new BigDecimal(1000000));
    }

    public BigDecimal getDurationNanoseconds() {
        return durationSeconds.multiply(new BigDecimal(1000000000));
    }

    public String toString() {
        BigDecimal valueOfOne = new BigDecimal(1);
        if (getDurationSeconds().compareTo(valueOfOne) >= 0) {
            return super.toString() + "\tpulse for " + getDurationSeconds() + "s";
        } else if (getDurationMilliseconds().compareTo(valueOfOne) >= 0) {
            return super.toString() + "\tpulse for " + getDurationMilliseconds() + "ms";
        } else if (getDurationMicroseconds().compareTo(valueOfOne) >= 0) {
            return super.toString() + "\tpulse for " + getDurationMicroseconds() + "us";
        } else {
            return super.toString() + "\tpulse for " + getDurationSeconds() + "us";
        }
    }

    // Generated getters and setters
    public PlotPointRaw getStartPulse() {
        return startPulse;
    }

    public PlotPointRaw getEndPulse() {
        return endPulse;
    }

}

File: src/main/kotlin/plotpoint/PlotPointPulse.kt
package plotpoint

import java.math.BigDecimal

open class PlotPointPulse(private val startPulse: PlotPoint,
                          private val endPulse: PlotPoint): PlotPoint(startPulse.timestampSeconds, startPulse.voltValue) {

    private val durationSeconds: BigDecimal = endPulse.timestampSeconds.subtract(startPulse.timestampSeconds)

    fun getDurationMilliseconds(): BigDecimal {
        return durationSeconds.multiply(BigDecimal(1000))
    }

    fun getDurationMicroseconds(): BigDecimal {
        return durationSeconds.multiply(BigDecimal(1000000))
    }

    fun getDurationNanoseconds(): BigDecimal {
        return durationSeconds.multiply(BigDecimal(1000000000))
    }

    override fun toString(): String {
        val valueOfOne = BigDecimal(1)
        return if (timestampSeconds.compareTo(valueOfOne) >= 0) {
            super.toString() + "\tpulse for " + timestampSeconds + "s"
        } else if (getDurationMilliseconds().compareTo(valueOfOne) >= 0) {
            super.toString() + "\tpulse for " + getDurationMilliseconds() + "ms"
        } else if (getDurationMicroseconds().compareTo(valueOfOne) >= 0) {
            super.toString() + "\tpulse for " + getDurationMicroseconds() + "us"
        } else {
            super.toString() + "\tpulse for " + timestampSeconds + "us"
        }
    }

    // Generated getters and setters
    fun getStartPulse(): PlotPoint {
        return startPulse
    }

    fun getEndPulse(): PlotPoint {
        return endPulse
    }
}